Add `PROPERTY_HINT_INPUT_NAME` for use with `@export_custom` to allow using input actions

Dynamic-Pistol opened the pull request:
Based off https://github.com/godotengine/godot-proposals/discussions/7559, this adds a new property hint for use, which creates an enum suggestion for `String`s/`StringName`s alongside an option to allow `"show_builtin"` to show builtin actions, docs are also included.

Note: this is a remake of another pr due to issues with github and me, and the fact it originally had an annotation

* *Bugsquad edit, closes: https://github.com/godotengine/godot-proposals/issues/11202*

AThousandShips commented on doc/classes/@GlobalScope.xml:
```diff
@@ -2930,7 +2930,11 @@
 		</constant>
-		<constant name="PROPERTY_HINT_MAX" value="38" enum="PropertyHint">
+		<constant name="PROPERTY_HINT_INPUT_NAME" value="38" enum="PropertyHint">
+			Hints that a string property is an input action name, and can select any action name from the input map.
+			If the [code]"show_builtin"[/code] hint, then built-in input actions can be used.
```
```suggestion
			Hints that a string property is an input action name, and can select any action name from the input map. Using the [code]"show_builtin"[/code] hint, built-in input actions can be used.
```
Or similar

AThousandShips commented:
You seem to have changed a bunch of GDScript output files now, please restore them

Dynamic-Pistol commented:
> You seem to have changed a bunch of GDScript output files now, please restore them

weird, i did initally as somehow the `utils.notest.gd` file was screwed,and while generating the tests, they got renewed, but i reverted the commit, so they shouldn't be merged? look at changed files tab

AThousandShips commented:
You removed the revert in your latest push for some reason, you probably have an out-of-date version of the branch locally

Also, in the future, make sure to create a new branch for each PR, you're contributing from `master` which is error prone

Dynamic-Pistol commented:
> You removed the revert in your latest push for some reason, you probably have an out-of-date version of the branch locally

Should be fixed now

> Also, in the future, make sure to create a new branch for each PR, you're contributing from `master` which is error prone

Got it

Dynamic-Pistol commented:
@AThousandShips for some reason, the macos build failed unit tests with `utils.notest.gd` even after i fixed it?

AThousandShips commented:
Not sure, you might not have registered the hint correctly

AThousandShips commented:
This is probably because the new constant is out of order, as `PROPERTY_HINT_LAYERS_AVOIDANCE` isn't exposed, meaning that the enum is missing an entry probably, though I'm not sure

Dynamic-Pistol commented:
> This is probably because the new constant is out of order, as `PROPERTY_HINT_LAYERS_AVOIDANCE` isn't exposed, meaning that the enum is missing an entry probably, though I'm not sure

So how would i fix it?

AThousandShips commented:
I don't know, try copying the enum from `utils.notest.gd` and open it in the editor and see if it spits out any errors and go from there

AThousandShips commented on modules/gdscript/tests/scripts/utils.notest.gd:
```diff
@@ -198,6 +197,10 @@ static func get_property_hint_name(hint: PropertyHint) -> String:
 		PROPERTY_HINT_PASSWORD:
 			return "PROPERTY_HINT_PASSWORD"
+ 		PROPERTY_HINT_INPUT_NAME:
+			return "PROPERTY_HINT_INPUT_NAME"
```
Duplicate, remove

AThousandShips commented on modules/gdscript/tests/scripts/utils.notest.gd:
```diff
@@ -253,4 +256,4 @@ static func get_property_usage_string(usage: int) -> String:
 		return "<invalid usage flags>"
 
-	return result.left(-1)
+	return result.left(-1)
```
```suggestion
	return result.left(-1)

```
Restore

AThousandShips commented on modules/gdscript/tests/scripts/utils.notest.gd:
All these format changes are invalid, please restore them

Dynamic-Pistol replied:
you mean revert the changes you told me?

AThousandShips replied:
What changes did I tell you? I mean the changes you added with spacing, they are incorrect

AThousandShips commented:
You've made merge commits, you need to push with `git push --force` or you'll merge the changes

I suggest squashing your changes as you have a lot of unnecessary commits in this PR, and you need to fix the branching issues you're having and fix the formatting

AThousandShips commented on modules/gdscript/tests/scripts/utils.notest.gd:
```diff
@@ -198,6 +198,8 @@ static func get_property_hint_name(hint: PropertyHint) -> String:
 		PROPERTY_HINT_PASSWORD:
 			return "PROPERTY_HINT_PASSWORD"
+		PROPERTY_HINT_INPUT_NAME:
+		    	return "PROPERTY_HINT_INPUT_NAME"
```
```suggestion
			return "PROPERTY_HINT_INPUT_NAME"
```

dalexeev commented on modules/gdscript/tests/scripts/utils.notest.gd:
```diff
@@ -254,3 +256,4 @@ static func get_property_usage_string(usage: int) -> String:
 		return "<invalid usage flags>"
 
 	return result.left(-1)
+
```
Unnecessary blank line at end of file. There must be exactly one newline (`\n`), not zero or two.

radiantgurl reviewed (changes requested):
![image](https://github.com/user-attachments/assets/80d19e70-9fab-429f-8adf-719822bc774e)
Works mostly as expected apart from a nullable action string. We should have it require an action key depending on the string hint.

For example if the string hint is `"default_empty"`, we should allow the empty option only in that case.

radiantgurl reviewed (approved):
Works as expected on 4.3.1 branch
![image](https://github.com/user-attachments/assets/1a5ad40a-7a35-4016-b167-998e70d5607b)

aaronfranke commented on doc/classes/@GlobalScope.xml:
```diff
@@ -2933,7 +2933,10 @@
 		</constant>
-		<constant name="PROPERTY_HINT_MAX" value="39" enum="PropertyHint">
+		<constant name="PROPERTY_HINT_INPUT_NAME" value="39" enum="PropertyHint">
+			Hints that a string property is an input action name, and can select any action name from the input map. Using the [code]"show_builtin"[/code] hint, built-in input actions can be used. Using the [code]"default_empty"[/code] hint, loose mode is enabled.
```
"default_empty" is a really bad name for a feature that allows entering custom values. I would rename it to something like "allow_custom" or similar. Internally Godot calls this `loose_mode`, which would work too.

dalexeev commented on core/object/object.h:
```diff
@@ -87,6 +87,7 @@ enum PropertyHint {
 	PROPERTY_HINT_PASSWORD,
 	PROPERTY_HINT_LAYERS_AVOIDANCE,
 	PROPERTY_HINT_DICTIONARY_TYPE,
+	PROPERTY_HINT_INPUT_NAME,
 	PROPERTY_HINT_TOOL_BUTTON,
```
```suggestion
	PROPERTY_HINT_TOOL_BUTTON,
	PROPERTY_HINT_INPUT_NAME,
```

Of course, until 4.4 we have the right to add constants in any order as long as it does not break compatibility with 4.3, but I would prefer to keep the chronological order.

Mickeon commented on doc/classes/@GlobalScope.xml:
```diff
@@ -2954,6 +2954,9 @@
 			Hints that a property will be changed on its own after setting, such as [member AudioStreamPlayer.playing] or [member GPUParticles3D.emitting].
 		</constant>
+		<constant name="PROPERTY_HINT_INPUT_NAME" value="41" enum="PropertyHint">
+			Hints that a string property is an input action name, and can select any action name from the input map. Using the [code]"show_builtin"[/code] hint, built-in input actions can be used. Using the [code]"loose_mode"[/code] hint, loose mode is enabled.
```
```suggestion
			Hints that a string property is the name of an input action. This allows the selection of any action name from the Input Map in the Project Settings. The hint string may contain two options separated by commas:
			- If it contains [code]"show_builtin"[/code], built-in input actions are included in the selection.
			- If it contains [code]"loose_mode"[/code], loose mode is enabled. This allows inserting any action name even if it's not present in the input map.
```

Mickeon commented on editor/editor_properties.cpp:
```diff
@@ -3754,6 +3780,31 @@ EditorProperty *EditorInspectorDefaultPlugin::get_editor_for_property(Object *p_
 				editor->setup(options, true, (p_hint == PROPERTY_HINT_ENUM_SUGGESTION));
 				return editor;
+			} else if (p_hint == PROPERTY_HINT_INPUT_NAME) {
+				EditorPropertyTextEnum *editor = memnew(EditorPropertyTextEnum);//see line 3606
```
You really shouldn't write line numbers in comments. They're very, very unreliable.

Mickeon commented on editor/editor_properties.cpp:
```diff
@@ -3602,6 +3603,31 @@ EditorProperty *EditorInspectorDefaultPlugin::get_editor_for_property(Object *p_
 				editor->setup(options, false, (p_hint == PROPERTY_HINT_ENUM_SUGGESTION));
 				return editor;
+			} else if (p_hint == PROPERTY_HINT_INPUT_NAME) {
+				EditorPropertyTextEnum *editor = memnew(EditorPropertyTextEnum);//should probably use a better editor gui with a searchbar
```
Given what you've expressed on [RocketChat](https://chat.godotengine.org/channel/input?msg=NsbTKSbSh7oFGnenX), this should at least be a TODO:
```suggestion
				// TODO: Should probably use a better editor GUI with a search bar.
				EditorPropertyTextEnum *editor = memnew(EditorPropertyTextEnum);
```

KoBeWi commented on editor/editor_properties.cpp:
```diff
@@ -3600,6 +3601,32 @@ EditorProperty *EditorInspectorDefaultPlugin::get_editor_for_property(Object *p_
+
+					String action_name = pi.name.substr(pi.name.find("/") + 1);
+					if (InputMap::get_singleton()->get_builtins().has(action_name)) {
+						if (hints.has("show_builtin")) {
```
You should cache this in a bool variable before the loop.
```C++
bool show_built_in = hints.has("show_builtin");
```

KoBeWi commented on editor/editor_properties.cpp:
```diff
@@ -3600,6 +3601,32 @@ EditorProperty *EditorInspectorDefaultPlugin::get_editor_for_property(Object *p_
+						continue;
+					}
+
+					String action_name = pi.name.substr(pi.name.find("/") + 1);
```
```suggestion
					const String action_name = pi.name.get_slicec('/', 1);
```

KoBeWi commented on editor/editor_properties.cpp:
```diff
@@ -3600,6 +3601,32 @@ EditorProperty *EditorInspectorDefaultPlugin::get_editor_for_property(Object *p_
+					}
+
+					String action_name = pi.name.substr(pi.name.find("/") + 1);
+					if (InputMap::get_singleton()->get_builtins().has(action_name)) {
```
`get_builtins()` should also be called once before the loop.

KoBeWi commented on editor/editor_properties.cpp:
```diff
@@ -3600,6 +3601,32 @@ EditorProperty *EditorInspectorDefaultPlugin::get_editor_for_property(Object *p_
+						if (hints.has("show_builtin")) {
+							builtin_options.append(action_name);
+						}
+						continue;
+					}
+					options.append(action_name);
```
```suggestion
					} else {
						options.append(action_name);
					}
```

KoBeWi commented on editor/editor_properties.cpp:
```diff
@@ -3752,6 +3779,32 @@ EditorProperty *EditorInspectorDefaultPlugin::get_editor_for_property(Object *p_
 				editor->setup(options, true, (p_hint == PROPERTY_HINT_ENUM_SUGGESTION));
 				return editor;
+			} else if (p_hint == PROPERTY_HINT_INPUT_NAME) {
+				// TODO: Should probably use a better editor GUI with a search bar.
+				EditorPropertyTextEnum *editor = memnew(EditorPropertyTextEnum);
+				Vector<String> options;
+				Vector<String> builtin_options;
+				List<PropertyInfo> pinfo;
+				ProjectSettings::get_singleton()->get_property_list(&pinfo);
+				Vector<String> hints = p_hint_text.split(",", false);
+
+				for (const PropertyInfo &pi : pinfo) {
+					if (!pi.name.begins_with("input/")) {
+						continue;
+					}
+
+					String action_name = pi.name.substr(pi.name.find("/") + 1);
+					if (InputMap::get_singleton()->get_builtins().has(action_name)) {
+						if (hints.has("show_builtin")) {
+							builtin_options.append(action_name);
+						}
+						continue;
+					}
+					options.append(action_name);
+				}
+				options.append_array(builtin_options);
+				editor->setup(options, false, hints.has("loose_mode"));
+				return editor;
```
Yes.

Dynamic-Pistol committed:
Add PROPERTY_HINT_INPUT_NAME

Added PROPERTY_HINT_INPUT_NAME for StringName based off https://github.com/godotengine/godot-proposals/discussions/7559

Calinou reviewed (commented):
Tested locally, it works as expected.

**Testing project:** [test_pr_96611.zip](https://github.com/user-attachments/files/19398281/test_pr_96611.zip)

Some feedback:

- If you don't set a default value (and the property is `null`, no dropdown appears to select a custom input action you created (or built-in actions if `show_builtin` is used):

![image](https://github.com/user-attachments/assets/f5203adb-6328-454f-a4dc-1f8930f7efe1)

This can only be done if you assign any default value such as an empty string (or a `: String`/`: StringName` type hint, as it's the default value):

![image](https://github.com/user-attachments/assets/d0832649-70ac-4149-99db-bbf525366640)

- If you create an exported input action property, the property has a default value assigned *and* the project has no custom properties, the property's height will be very small (lower than all other property types):

![image](https://github.com/user-attachments/assets/509254b0-c71f-40dd-91e0-2a8f92a4897b)

This does not occur with the default value of `null`. It's probably not a big issue as I don't think it'll be encountered often anyway.

- Having the hint string separated by a space (`loose_mode, show_builtin` instead of `loose_mode,show_builtin`) causes it to break:

![image](https://github.com/user-attachments/assets/3a33067a-f516-4303-919d-28fc6a30fde2)

This can be fixed by stripping spaces when reading each hint, so that leading/trailing whitespace doesn't matter.

- Having a typo in the hint string (e.g. `show_builti` instead of `show_builtin`) won't print any errors to the editor. This should be added if possible.

Dynamic-Pistol commented:
> * If you don't set a default value (and the property is `null`, no dropdown appears to select a custom input action you created (or built-in actions if `show_builtin` is used):

This seems to be an issue with `EditorPropertyTextEnum` (it also happens with it), it seems to have to do with the object being variant with a value of null, only solution i can think of is to set the object to either the first input value or an empty string, but i don't feel like that's right, and i could try to fix the issue withe the text enum prop editor, but it seems out of scope for this pr, so either the solution i mentioned (if possible) or i just let it be someone else's problem

> * If you create an exported input action property, the property has a default value assigned _and_ the project has no custom properties, the property's height will be very small (lower than all other property types):

"the project has no custom properties" uhh what custom properties? like no input actions?  be more clear, also sounds like another  `EditorPropertyTextEnum` issue, not sure if you can remake it with `XXX_ENUM_SUGGESTION` but it still would be the source problem, again this goes out of scope for this pr


> * Having the hint string separated by a space (`loose_mode, show_builtin` instead of `loose_mode,show_builtin`) causes it to break:

Okay, it should be fixed with a simple
`Vector<String> hints = p_hint_text.remove_chars(" ").split(",", false);`

> * Having a typo in the hint string (e.g. `show_builti` instead of `show_builtin`) won't print any errors to the editor. This should be added if possible.

I don't think this is feasible with a property hint, none of the hints have errors when you use `@export_custom`, i suppose this could work

```c++
for (const String& hint : hints) {
        if (hint != "show_builtin" || hint != "loose_mode") {
            ERR_FAIL_V_MSG(get_input_action_editor(""), "Invalid PROPERTY_HINT_INPUT_NAME with hint \"%s\"");
        }
    }
```

but i want you to decide with others if this is really what is needed

---

Okay this should be done, @Calinou there shouldn't be any issues, i will push these changes tomorrow since i am going to sleep at the time this is written

some others changes include

- Adding another comment for future ref
- Adding a `is_string_name` parameter because in the old code the 2nd parameter of setup was always false for some reason

Have a nice day

Calinou commented:
> "the project has no custom properties" uhh what custom properties?

Yes, I meant custom input actions defined in the input map.

AThousandShips commented on editor/editor_properties.cpp:
```diff
@@ -3534,6 +3535,45 @@ static EditorPropertyRangeHint _parse_range_hint(PropertyHint p_hint, const Stri
+	Vector<String> builtin_options;
+	List<PropertyInfo> pinfo;
+	ProjectSettings::get_singleton()->get_property_list(&pinfo);
+	Vector<String> hints = p_hint_text.remove_chars(" ").split(",", false);
```
```suggestion
	Vector<String> hints = p_hint_text.remove_char(' ').split(",", false);
```

AThousandShips commented on core/object/object.h:
```diff
@@ -94,6 +94,7 @@ enum PropertyHint {
 	PROPERTY_HINT_ONESHOT, ///< the property will be changed by self after setting, such as AudioStreamPlayer.playing, Particles.emitting.
 	PROPERTY_HINT_NO_NODEPATH, /// < this property will not contain a NodePath, regardless of type (Array, Dictionary, List, etc.). Needed for SceneTreeDock.
 	PROPERTY_HINT_GROUP_ENABLE, ///< used to make the property's group checkable. Only use for boolean types.
+	PROPERTY_HINT_INPUT_NAME,
```
Needs a comment to document it

Dynamic-Pistol replied:
Will get some suggestions from the rocketchat instance

KoBeWi replied:
Discussion for reference: https://chat.godotengine.org/channel/editor?msg=z52j43s2oMhGp6bhJ

AThousandShips commented on editor/editor_properties.cpp:
```diff
@@ -3570,6 +3571,45 @@ static EditorPropertyRangeHint _parse_range_hint(PropertyHint p_hint, const Stri
+	ProjectSettings::get_singleton()->get_property_list(&pinfo);
+	Vector<String> hints = p_hint_text.remove_char(' ').split(",", false);
+
+	//Code that shows an error when an invalid hint is entered
+	//for (const String& hint : hints) {
+	//    if (hint != "show_builtin" || hint != "loose_mode") {
+	//        ERR_FAIL_V_MSG(get_input_action_editor("", is_string_name), vformat("Invalid PROPERTY_HINT_INPUT_NAME with hint \"%s\"", hint));
+	//    }
+	//}
```
This should either be removed or uncommented IMO

Dynamic-Pistol replied:
It's for future contributors, Callinou asked for errors but none of the other property hints error, so incase it changes in the future, this would just need to be uncommented

AThousandShips replied:
Should probably have a note to that effect

Mickeon replied:
You can potentially slap a TODO comment if this is not strictly necessary to have it merged now.
```suggestion
	// TODO: Show an error when an invalid hint is entered.
```

Dynamic-Pistol replied:
@Mickeon the whole point of the comment is so that when others implement errors for the other hints, they can just uncomment this and it will be done, but if you think it's too useless or something, then i can just remove it

theraot replied:
@Dynamic-Pistol  I'd say remove it.

---

For what I gather:

1. That is a desired feature, but it is beyond the scope of this one.
2. You do not want to lose it.

You say it is for future contributors, but if it is done it is done, you are the future contributor. *In fact, if somebody implemented it by just uncommenting it, you'd deserve - at least - co-authorship.*

So, this is my suggestion: you can make a new PR for it, say in the description that it depends on this one. And you can point to the comments where it was discussed to justify it (assuming there is no proposal for it).

Unless you really don't care about it. In that case, just remove it. If or when there is a proposal, the code can be salvaged from this conversation. *In a sense, it is already a comment, since it is here in github*, so you don't need to keep in the code. *Anyway, if there is no proposal, I think it would be better off with one, that way it would make easier to find it and track it.*

Calinou reviewed (commented):
Tested locally, it works as expected. The UX is good to me.

Repiteo commented:
Thanks! Congratulations on your first merged contribution! 🎉
